airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/AutoDisableConnectionActivityImpl.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.workers.temporal.scheduling.activities;
+
+import static io.airbyte.scheduler.models.Job.REPLICATION_TYPES;
+
+import io.airbyte.commons.features.FeatureFlags;
+import io.airbyte.config.Configs;
+import io.airbyte.config.StandardSync;
+import io.airbyte.config.StandardSync.Status;
+import io.airbyte.config.persistence.ConfigRepository;
+import io.airbyte.scheduler.models.Job;
+import io.airbyte.scheduler.models.JobStatus;
+import io.airbyte.scheduler.persistence.JobPersistence;
+import io.airbyte.workers.temporal.exception.RetryableException;
+import java.time.temporal.ChronoUnit;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import lombok.AllArgsConstructor;
+
+@AllArgsConstructor
+public class AutoDisableConnectionActivityImpl implements AutoDisableConnectionActivity {
+
+  private ConfigRepository configRepository;
+  private JobPersistence jobPersistence;
+  private FeatureFlags featureFlags;
+  private Configs configs;
+
+  // Given a connection id and current timestamp, this activity will set a connection to INACTIVE if
+  // either:
+  // - fails jobs consecutively and hits the `configs.getMaxFailedJobsInARowBeforeConnectionDisable()`
+  // limit
+  // - all the jobs in the past `configs.getMaxDaysOfOnlyFailedJobsBeforeConnectionDisable()` days are
+  // failures,
+  // and that the connection's first job is at least that many days old
+  @Override
+  public void autoDisableFailingConnection(final AutoDisableConnectionActivityInput input) {
+    if (featureFlags.autoDisablesFailingConnections()) {
+      try {
+        final int maxDaysOfOnlyFailedJobs = configs.getMaxDaysOfOnlyFailedJobsBeforeConnectionDisable();
+        final List<JobStatus> jobStatuses = jobPersistence.listJobStatusWithConnection(input.getConnectionId(), REPLICATION_TYPES,
+            input.getCurrTimestamp().minus(maxDaysOfOnlyFailedJobs, ChronoUnit.DAYS));
+
+        if (jobStatuses.size() == 0)
+          return;
+
+        int numFailures = 0;
+
+        // jobs are sorted from most recent to least recent
+        for (final JobStatus jobStatus : jobStatuses) {
+          if (jobStatus == JobStatus.FAILED) {
+            numFailures++;
+            if (numFailures == configs.getMaxFailedJobsInARowBeforeConnectionDisable())
+              break;
+          } else if (jobStatus == JobStatus.SUCCEEDED) {
+            return;
+          }
+        }
+
+        // if failures are under the max failed limit but are the only jobs in the last max days limit,
+        // check to make sure that the connection is at least that many days old (based from created at
+        // of first job)
+        if (numFailures != configs.getMaxFailedJobsInARowBeforeConnectionDisable()) {
+          final Optional<Job> optionalFirstJob = jobPersistence.getFirstReplicationJob(input.getConnectionId());
+          if (optionalFirstJob.isPresent()) {
```

Comment:
Is it gonna be empty for the first run? Here if it is empty, we still de-activate the connection which doesn't sounds right.

Reply:
It should only be empty if there were only cancelled jobs or there were no job runs for that connection. If the activity is run when there are no jobs previously, it will not disable the connection because of the earlier line that checks if there have been any runs in the past 14 days:
```
         final List<JobStatus> jobStatuses = jobPersistence.listJobStatusWithConnection(input.getConnectionId(), REPLICATION_TYPES,
            input.getCurrTimestamp().minus(maxDaysOfOnlyFailedJobs, ChronoUnit.DAYS));
        if (jobStatuses.size() == 0)
          return;
```

Reply:
Won't jobStatuses contains contain the current failed job? If so we won't return

Reply:
we talked through this online. came to the conclusion that the option first job should not be empty on the first run, it should in fact contain that first run.